Crash Report: android - v1.5.8 (android)

## Crash Report

**Version:** 1.5.8
**Platform:** android
**Device:** android localhost


### Crash Log
```
=== JAVA CRASH REPORT ===
Time: 2026-03-21 09:47:57
Thread: QtBluetoothLEHandlerThread
Exception: android.os.DeadSystemRuntimeException
Message: android.os.DeadSystemException

=== STACK TRACE ===
android.os.DeadSystemRuntimeException: android.os.DeadSystemException
	at android.bluetooth.BluetoothGatt.writeCharacteristic(BluetoothGatt.java:1501)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE.executeWriteJob(QtBluetoothLE.java:1699)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE.performNextIO(QtBluetoothLE.java:1554)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE.access$1600(QtBluetoothLE.java:42)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE$3.run(QtBluetoothLE.java:1489)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: android.os.DeadSystemException
	... 10 more

=== DEVICE INFO ===
Android: 14 (API 34)
Device: Teclast M50Mini

```

### Debug Log (last lines)
```
[2520.807] DEBUG [Memory] RSS: 87.3 MB  Java heap: 6.7 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2552.774] DEBUG [Screensaver] Video transition #11 RSS:87.3 MB
[2580.799] DEBUG BatteryManager: battery= 95 % mode= On(55-65%) charger= OFF discharging= true status= DISCHARGING plugged= UNPLUGGED
[2580.809] DEBUG [Memory] RSS: 86.9 MB  Java heap: 7.1 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2590.280] DEBUG [Screensaver] Video transition #12 RSS:86.9 MB
[2627.819] DEBUG [Screensaver] Video transition #13 RSS:86.9 MB
[2640.808] DEBUG [Memory] RSS: 86.6 MB  Java heap: 7.5 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2665.350] DEBUG [Screensaver] Video transition #14 RSS:86.6 MB
[2700.808] DEBUG [Memory] RSS: 87.1 MB  Java heap: 7.8 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2733.680] DEBUG [Screensaver] Video transition #15 RSS:87.1 MB
[2760.806] DEBUG [Memory] RSS: 86.1 MB  Java heap: 8.1 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2820.806] DEBUG [Memory] RSS: 86.5 MB  Java heap: 8.4 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2850.175] DEBUG [Screensaver] Video transition #16 RSS:86.5 MB
[2880.326] DEBUG [Screensaver] Video transition #17 RSS:86.5 MB
[2880.799] DEBUG BatteryManager: battery= 95 % mode= On(55-65%) charger= OFF discharging= true status= DISCHARGING plugged= UNPLUGGED
[2880.809] DEBUG [Memory] RSS: 87.2 MB  Java heap: 8.8 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2940.808] DEBUG [Memory] RSS: 86.1 MB  Java heap: 9.2 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[2955.964] DEBUG [Screensaver] Video transition #18 RSS:86.1 MB
[3000.808] DEBUG [Memory] RSS: 87.7 MB  Java heap: 9.5 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3016.548] DEBUG [Screensaver] Video transition #19 RSS:87.7 MB
[3060.808] DEBUG [Memory] RSS: 87.3 MB  Java heap: 9.8 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3084.870] DEBUG [Screensaver] Video transition #20 RSS:87.3 MB
[3120.806] DEBUG [Memory] RSS: 85.8 MB  Java heap: 10.1 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3124.927] DEBUG [Screensaver] Video transition #21 RSS:85.8 MB
[3162.378] DEBUG [Screensaver] Video transition #22 RSS:85.8 MB
[3180.799] DEBUG BatteryManager: battery= 94 % mode= On(55-65%) charger= OFF discharging= true status= DISCHARGING plugged= UNPLUGGED
[3180.806] DEBUG [Memory] RSS: 87.2 MB  Java heap: 10.5 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3204.412] DEBUG [Screensaver] Video transition #23 RSS:87.2 MB
[3240.808] DEBUG [Memory] RSS: 86.1 MB  Java heap: 10.8 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3266.824] DEBUG [Screensaver] Video transition #24 RSS:86.1 MB
[3299.870] DEBUG [Screensaver] Video transition #25 RSS:86.1 MB
[3300.807] DEBUG [Memory] RSS: 86.5 MB  Java heap: 11.1 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3348.307] DEBUG [Screensaver] Video transition #26 RSS:86.5 MB
[3360.808] DEBUG [Memory] RSS: 88.4 MB  Java heap: 11.5 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3392.312] DEBUG [Screensaver] Video transition #27 RSS:88.4 MB
[3420.809] DEBUG [Memory] RSS: 87.0 MB  Java heap: 11.9 / 256.0 MB  QObjects: 1488  peak: 120.5 MB
[3421.387] WARN  makeCurrent(): no EGLSurface, likely Surface destroyed by Android.
[3421.387] WARN  QRhiGles2: Failed to make context current. Expect bad things to happen.
[3421.388] WARN  Failed to start frame
[3421.423] DEBUG App suspended - sleeping scale (DE1 stays awake)
[3421.424] DEBUG BatteryManager: Ensuring charger is ON (app exit/suspend safety)
[3421.425] DEBUG Application exiting - shutting down devices
[3421.429] DEBUG Sending DE1 to sleep on app exit
[3421.430] DEBUG Waiting for BLE queue to drain before exit...
[3423.332] WARN  BLE queue drain timed out after 2000 ms — sleep command may not have been delivered.
[3423.332] DEBUG BatteryManager: Ensuring charger is ON (app exit/suspend safety)
[3423.334] DEBUG AccessibilityManager shutting down

========== SESSION START: 2026-03-22T07:42:11 ==========
[   0.161] DEBUG App started - version 1.5.8 build 3176 (release) built Mar 19 2026 19:01:33 at "2026-03-22T07:42:11"
```

---
*Auto-reported by Decenza DE1 app*

--- a/android/src/io/github/kulitorum/decenza_de1/DecenzaActivity.java
+++ b/android/src/io/github/kulitorum/decenza_de1/DecenzaActivity.java
@@ -80,15 +80,17 @@ public class DecenzaActivity extends QtActivity {
         Thread.setDefaultUncaughtExceptionHandler((thread, throwable) -> {
             // DeadSystemException means Android's Bluetooth system service died
             // (deep sleep, OEM power management, system_server crash).
-            // If it happened on the BLE thread, write a flag file so the C++ side
-            // can detect it and trigger BLE recovery (disconnect + reconnect).
+            // Create a flag file so the C++ side can detect it and trigger BLE
+            // recovery (disconnect + reconnect). This fires on any thread — we
+            // don't filter by thread name since DeadSystemException only occurs
+            // when Android system services die, which always affects BLE.
             // Don't crash the app — the rest of the app is fine, only BLE is dead.
             if (isDeadSystemException(throwable)) {
                 Log.w(TAG, "DeadSystemException on thread " + thread.getName()
                         + " — Bluetooth system died, signaling BLE recovery");
                 try {
                     File flagFile = new File(getFilesDir(), "ble_dead_system");
-                    new FileWriter(flagFile).close();
+                    flagFile.createNewFile();
                     Log.w(TAG, "Wrote BLE recovery flag: " + flagFile.getAbsolutePath());
                 } catch (Exception e) {
                     Log.e(TAG, "Failed to write BLE recovery flag: " + e.getMessage());

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -707,9 +707,8 @@ int main(int argc, char *argv[])
     auto* bleRecoveryTimer = new QTimer();
     bleRecoveryTimer->setInterval(10000);
     QObject::connect(bleRecoveryTimer, &QTimer::timeout,
-                     [&bleManager, &de1Device, &settings]() {
-        // Java getFilesDir() = /data/data/<pkg>/files/ on Android
-        // QStandardPaths::AppDataLocation = /data/data/<pkg>/files/ on Android
+                     [&bleManager, &de1Device]() {
+        // Use same path as CrashHandler (proven to match Java getFilesDir())
         QString flagPath = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation)
                          + "/ble_dead_system";
         QFile flagFile(flagPath);

